spring-data-geode/src/test/java/org/springframework/data/gemfire/function/execution/onservers/FunctionsReturnResultsFromAllServersIntegrationTests.java

```diff
@@ -0,0 +1,172 @@
+				String.format("-D%s=%d", GEMFIRE_CACHE_SERVER_PORT_PROPERTY, PORT_2));
+
+		waitForServerToStart(DEFAULT_HOSTNAME, PORT_2);
+	}
+
+	@ClientCacheApplication(servers = {@ClientCacheApplication.Server(port = PORT_1), @ClientCacheApplication.Server(port = PORT_2)})
+	@Configuration
+	@EnableGemfireFunctionExecutions(basePackageClasses = AllServersAdminFunctions.class)
+	@EnableGemfireCacheTransactions
+	static class Config { }
+
+	@AfterClass
+	public static void stopGemFireServer() {
+		stop(gemfireServer1);
+		stop(gemfireServer2);
+	}
+
+	@Test
+	public void executeFunctionOnAllServers() {
+		List<List<Metric>> metrics = allServersAdminFunctions.getAllMetrics();
+		assertThat(metrics.size()).isEqualTo(2);
+	}
+
+	@Test
+	public void executeFunctionOnSingleServer() {
+		List<Metric> metrics = singleServerAdminFunctions.getAllMetrics();
+		assertThat(metrics.size()).isEqualTo(672);
+	}
+
+	static class MetricsFunctionServerProcess {
+
+		private static final int DEFAULT_CACHE_SERVER_PORT = 40404;
+
+		private static final String CACHE_SERVER_PORT_PROPERTY = "spring.data.gemfire.cache.server.port";
+		private static final String GEMFIRE_LOG_LEVEL = "error";
+		private static final String GEMFIRE_NAME = "MetricsServer" + getCacheServerPort();
+
+		public static void main(String[] args) throws Exception {
+			registerFunctions(startCacheServer(newGemFireCache()));
+		}
+
+		private static Cache newGemFireCache() {
+
+			return new CacheFactory()
+					.set("name", GEMFIRE_NAME)
+					.set("log-level", GEMFIRE_LOG_LEVEL)
+					.create();
+		}
+
+		private static Cache startCacheServer(Cache gemfireCache) throws IOException {
+			CacheServer cacheServer = gemfireCache.addCacheServer();
+			cacheServer.setPort(getCacheServerPort());
+			cacheServer.start();
+			return gemfireCache;
+		}
+
+		private static int getCacheServerPort() {
+			return Integer.getInteger(CACHE_SERVER_PORT_PROPERTY, DEFAULT_CACHE_SERVER_PORT);
+		}
+
+		private static Cache registerFunctions(Cache gemfireCache) {
+
+			FunctionService.registerFunction(new GetAllMetricsFunction());
+
+			return gemfireCache;
+		}
+	}
+
+	static class GetAllMetricsFunction implements Function<List<Metric>> {
+
+		private final InternalDistributedSystem system =
+				(InternalDistributedSystem) CacheFactory.getAnyInstance().getDistributedSystem();
+
+		@Override
+		public void execute(FunctionContext context) {
+			List<Metric> allMetrics = new ArrayList<>();
+			StatisticsManager statisticsManager = system.getStatisticsManager();
+			for (Statistics statistics : statisticsManager.getStatsList()) {
+				StatisticsType statisticsType = statistics.getType();
+				for (StatisticDescriptor descriptor : statisticsType.getStatistics()) {
+					String statName = descriptor.getName();
+					Metric metric = new Metric(statName, statistics.get(statName), statisticsType.getName(), statistics.getTextId());
+					allMetrics.add(metric);
+				}
+			}
+			context.getResultSender().lastResult(allMetrics);
```

Comment:
While I like your idea of mimicking the users use case, we can simplify this to something like...

`context.getResultSender.lastResult(new Metric("serverOneMetric", 1, "CPU", "System"));`

And then, for Server 2:

`context.getResultSender.lastResul(new Metric("serverTwoMetric", 1024, "Memory", "System"));`

Then our test cases could assert 2 Metrics in the all servers test case and 1 Metric in the 1 server test case.

Reply:
Simplified, now sends back a list of 10 made up Metrics instead of real metrics.